### PR TITLE
[RDY] vim-patch:7.4.502

### PR DESCRIPTION
**Pull request by glts**
```
Problem:    Language mapping also applies to mapped characters.
Solution:   Add the 'langnoremap' option, when on 'langmap' does not apply to
            mapped characters. (Christian Brabandt)
```

Straightforward port, no difficulties.

Original patch:

``` diff
diff --git a/runtime/doc/options.txt b/runtime/doc/options.txt
--- a/runtime/doc/options.txt
+++ b/runtime/doc/options.txt
@@ -4589,6 +4589,18 @@ A jump table for the options with a shor
        :source $VIMRUNTIME/menu.vim
 <  Warning: This deletes all menus that you defined yourself!

+                   *'langnoremap'* *'lnr'*
+'langnoremap' 'lnr'    boolean (default off)
+           global
+           {not in Vi}
+           {only available when compiled with the |+langmap|
+           feature}
+   When on, setting 'langmap' does not apply to characters resulting from
+   a mapping.  This basically means, if you noticed that setting
+   'langmap' disables some of your mappings, try setting this option.
+   This option defaults to off for backwards compatibility.  Set it on if
+   that works for you to avoid mappings to break.
+
                    *'laststatus'* *'ls'*
 'laststatus' 'ls'  number  (default 1)
            global
diff --git a/runtime/vimrc_example.vim b/runtime/vimrc_example.vim
--- a/runtime/vimrc_example.vim
+++ b/runtime/vimrc_example.vim
@@ -1,7 +1,7 @@
 " An example for a vimrc file.
 "
 " Maintainer:  Bram Moolenaar <Bram@vim.org>
-" Last change: 2014 Feb 05
+" Last change: 2014 Nov 05
 "
 " To use it, copy it to
 "     for Unix and OS/2:  ~/.vimrc
@@ -95,3 +95,10 @@ if !exists(":DiffOrig")
   command DiffOrig vert new | set bt=nofile | r ++edit # | 0d_ | diffthis
          \ | wincmd p | diffthis
 endif
+
+if has('langmap') && exists('+langnoremap')
+  " Prevent that the langmap option applies to characters that result from a
+  " mapping.  If unset (default), this may break plugins (but it's backward
+  " compatible).
+  set langnoremap
+endif
diff --git a/src/nvim/macros.h b/src/nvim/macros.h
--- a/src/nvim/macros.h
+++ b/src/nvim/macros.h
@@ -128,13 +128,18 @@
  * Adjust chars in a language according to 'langmap' option.
  * NOTE that there is no noticeable overhead if 'langmap' is not set.
  * When set the overhead for characters < 256 is small.
- * Don't apply 'langmap' if the character comes from the Stuff buffer.
+ * Don't apply 'langmap' if the character comes from the Stuff buffer or from
+ * a mapping and the langnoremap option was set.
  * The do-while is just to ignore a ';' after the macro.
  */
 # ifdef FEAT_MBYTE
 #  define LANGMAP_ADJUST(c, condition) \
     do { \
-   if (*p_langmap && (condition) && !KeyStuffed && (c) >= 0) \
+   if (*p_langmap \
+       && (condition) \
+       && (!p_lnr || (p_lnr && typebuf_maplen() == 0)) \
+       && !KeyStuffed \
+       && (c) >= 0) \
    { \
        if ((c) < 256) \
        c = langmap_mapchar[c]; \
@@ -145,7 +150,11 @@
 # else
 #  define LANGMAP_ADJUST(c, condition) \
     do { \
-   if (*p_langmap && (condition) && !KeyStuffed && (c) >= 0 && (c) < 256) \
+   if (*p_langmap \
+       && (condition) \
+       && (!p_lnr || (p_lnr && typebuf_maplen() == 0)) \
+       && !KeyStuffed \
+       && (c) >= 0 && (c) < 256) \
        c = langmap_mapchar[c]; \
     } while (0)
 # endif
diff --git a/src/nvim/option.c b/src/nvim/option.c
--- a/src/nvim/option.c
+++ b/src/nvim/option.c
@@ -1691,6 +1691,13 @@ static struct vimoption
                (char_u *)NULL, PV_NONE,
 #endif
                {(char_u *)"", (char_u *)0L} SCRIPTID_INIT},
+    {"langnoremap",  "lnr",   P_BOOL|P_VI_DEF,
+#ifdef FEAT_LANGMAP
+               (char_u *)&p_lnr, PV_NONE,
+#else
+               (char_u *)NULL, PV_NONE,
+#endif
+               {(char_u *)FALSE, (char_u *)0L} SCRIPTID_INIT},
     {"laststatus",  "ls",   P_NUM|P_VI_DEF|P_RALL,
 #ifdef FEAT_WINDOWS
                (char_u *)&p_ls, PV_NONE,
diff --git a/src/nvim/option.h b/src/nvim/option.h
--- a/src/nvim/option.h
+++ b/src/nvim/option.h
@@ -576,6 +576,7 @@ EXTERN char_u   *p_kp;      /* 'keywordprg' */
 EXTERN char_u  *p_km;      /* 'keymodel' */
 #ifdef FEAT_LANGMAP
 EXTERN char_u  *p_langmap; /* 'langmap'*/
+EXTERN int p_lnr;      /* 'langnoremap' */
 #endif
 #if defined(FEAT_MENU) && defined(FEAT_MULTI_LANG)
 EXTERN char_u  *p_lm;      /* 'langmenu' */
diff --git a/src/nvim/version.c b/src/nvim/version.c
--- a/src/nvim/version.c
+++ b/src/nvim/version.c
@@ -742,6 +742,8 @@ static char *(features[]) =
 static int included_patches[] =
 {   /* Add new patch number below this line */
 /**/
+    502,
+/**/
     501,
 /**/
     500,
```

### CODE REVIEW

**Comment by glts** [comment]
Funnily enough, patch 7.4.552, which has already been ported in #1742, actually relies on this new `'langnoremap'` option [in legacy test `mapping_spec.lua`](https://github.com/neovim/neovim/blob/master/test/functional/legacy/mapping_spec.lua#L23). But that fails silently and passes because it doesn't test what they think it tests (also in upstream Vim) ...

File: runtime/doc/options.txt [hunk 1]
```diff
@@ -4351,6 +4351,18 @@ A jump table for the options with a short description can be found at |Q_op|.
 		:source $VIMRUNTIME/menu.vim
 <	Warning: This deletes all menus that you defined yourself!
 
+					*'langnoremap'* *'lnr'*
+'langnoremap' 'lnr'	boolean (default off)
+			global
+			{not in Vi}
+			{only available when compiled with the |+langmap|
```
Comment (Pyrohh) [inline runtime/doc/options.txt]: This can be removed, Nvim is always built with langmap.

File: runtime/vimrc_example.vim [hunk 2]
```diff
@@ -80,3 +80,10 @@ if !exists(":DiffOrig")
   command DiffOrig vert new | set bt=nofile | r ++edit # | 0d_ | diffthis
                  \ | wincmd p | diffthis
 endif
+
+if exists('+langnoremap')
```
Comment (Pyrohh) [inline runtime/vimrc_example.vim]: this can be removed, as it should always be true.

**Comment by glts** [comment]
@Pyrohh Thanks, done. I suppose it's ok to just continue pushing commits, and then rebase and squash when final approval is here, right?

**Comment by Pyrohh** [comment]
For small patches like these, it should be fine, but maybe not for larger ones.

**Comment by Pyrohh** [comment]
LGTM

File: runtime/vimrc_example.vim [hunk 3]
```diff
@@ -80,3 +80,8 @@ if !exists(":DiffOrig")
   command DiffOrig vert new | set bt=nofile | r ++edit # | 0d_ | diffthis
                  \ | wincmd p | diffthis
 endif
+
+" Prevent that the langmap option applies to characters that result from a
+" mapping.  If unset (default), this may break plugins (but it's backward
+" compatible).
+set langnoremap
```
Comment (justinmk) [inline runtime/vimrc_example.vim]: :+1:  glad to see you didn't add the `if has('langmap') && exists('+langnoremap')` check

Comment (glts) [inline runtime/vimrc_example.vim]: @justinmk I have to admit that @Pyrohh had a hand in this ...

**Comment by justinmk** [comment]
merged

**Pull request closed without merging**